### PR TITLE
logind: fix abnormal switching causing the screen to go black

### PR DESCRIPTION
**Pull request by wxphaha**
After logind receives the SIGRTMIN signal from the kernel, it will execute manager_vt_switch---session_leave_vt---session_device_pause_all,The device permissions of the session are removed here;under normal circumstances, the tty value read from /sys/class/tty/tty0/active changes and switchesto a new session,give the new session resume device permissions. But under abnormal circumstances (such as switching quickly on a device using wayland; and sometimes the kernel will suddenly send a SIGRTMIN signal, but nothing changes),In these cases, logind does not give session resume device permission, causing the device to have a black screen and suspended animation.

### CODE REVIEW

File: src/login/logind-seat.c [hunk 1]
```diff
@@ -225,8 +225,14 @@ int seat_set_active(Seat *s, Session *session) {
         assert(s);
         assert(!session || session->seat == s);
 
-        if (session == s->active)
+        if (session == s->active) {
+		if (session && session->leaved) {
```
Comment (YHNdnzj) [inline src/login/logind-seat.c]: indentation??

File: src/login/logind-seat.c [hunk 2]
```diff
@@ -225,8 +225,14 @@ int seat_set_active(Seat *s, Session *session) {
...
+			log_warning("seat_set_active something wrong resume cur");
```
Comment (YHNdnzj) [inline src/login/logind-seat.c]: Please switch to `log_debug` and rephrase the message.

File: src/login/logind-session.h [hunk 3]
```diff
@@ -106,6 +106,7 @@ struct Session {
         bool stopping:1;
 
         bool was_active:1;
+        bool leaved:1;
```
Comment (YHNdnzj) [inline src/login/logind-session.h]: left

File: src/login/logind-seat.c (hunk 1 above)
Comment (YHNdnzj) [reply src/login/logind-seat.c]: Please add some explanation here why this is needed.

File: src/login/logind-session.c [hunk 4]
```diff
@@ -1427,6 +1427,7 @@ void session_leave_vt(Session *s) {
                 return;
 
         session_device_pause_all(s);
+        s->leaved=true;
```
Comment (YHNdnzj) [inline src/login/logind-session.c]:
```suggestion
        s->leaved = true;
        
```

**Review by YHNdnzj — CHANGES_REQUESTED** [review]

File: src/login/logind-session.h (hunk 3 above)
Comment (wxphaha) [reply src/login/logind-session.h]: thank you for your reply. But I don’t understand the meaning of left here. Can you explain it further? please。

Comment (YHNdnzj) [reply src/login/logind-session.h]: "leaved" is not the correct word...

File: src/login/logind-seat.c [hunk 5]
```diff
@@ -225,8 +225,21 @@ int seat_set_active(Seat *s, Session *session) {
...
+         * this time, if the session is active and there is no change in the
+         * session, then the session does not have the permissions of the device,
+         * and the machine will have a black screen and suspended animation.
+         * Therefore, if the active session has executed session_leave_vt ,
+         * A resume is required here. */
+        if (session == s->active) {
+                if (session && session->left) {
+                        log_debug("Active but the session remains unchanged,resume the session that has been left.");
+                        session_device_resume_all(session);
+                        session->left = false;
```
Comment (poettering) [inline src/login/logind-seat.c]: hmm, why maintain the boolean? why not just call session_device_resume_all() unconditionally?

i mean, the call is a NOP anyway for devices already resumed. Hence it's always safe to just call this, and it will unpause everything that might be resume or be a NOP.

Comment (wxphaha) [reply src/login/logind-seat.c]: Thank you, I have made modifications according to your comments,please review

File: src/login/logind-seat.c [hunk 6]
```diff
@@ -225,8 +225,20 @@ int seat_set_active(Seat *s, Session *session) {
...
+        /* When logind receives the SIGRTMIN signal from the kernel, it will
+         * execute session_leave_vt and stop all devices of the session; at
+         * this time, if the session is active and there is no change in the
+         * session, then the session does not have the permissions of the device,
+         * and the machine will have a black screen and suspended animation.
+         * Therefore, if the active session has executed session_leave_vt ,
+         * A resume is required here. */
+        if (session == s->active) {
+                if (session) {
+                        log_debug("Active but the session remains unchanged,resume the session that maybe left.");
```
Comment (poettering) [inline src/login/logind-seat.c]: there's a space missing after the ","

**Comment by poettering** [comment]
lgtm

File: src/login/logind-seat.c (hunk 6 above)
Comment (YHNdnzj) [inline src/login/logind-seat.c]: Maybe rephrase this as 

```
Active session remains unchanged, resuming session devices.
```

**Review by bluca — APPROVED** [review]
